cstore_compression.c

```diff
@@ -0,0 +1,165 @@
+
+#endif
+
+
+
+/*
+ * CompressBuffer compresses the given buffer with the given compression type
+ * outputBuffer enlarged to contain compressed data. The function returns true
+ * if compression is done, returns false if compression is not done.
+ * outputBuffer is valid only if the function returns true.
+ */
+bool
+CompressBuffer(StringInfo inputBuffer, StringInfo outputBuffer,
+			   CompressionType compressionType)
+{
+	uint64 maximumLength = PGLZ_MAX_OUTPUT(inputBuffer->len) + CSTORE_COMPRESS_HDRSZ;
+	bool compressionResult = false;
+
+	if (compressionType != COMPRESSION_PG_LZ)
+	{
+		return false;
+	}
+
+	resetStringInfo(outputBuffer);
+	enlargeStringInfo(outputBuffer, maximumLength);
+
+#if PG_VERSION_NUM >= 90500
+	int32 compressedByteCount = 0;
+
+	compressedByteCount = pglz_compress((const char *) inputBuffer->data,
+										inputBuffer->len,
+										CSTORE_COMPRESS_RAWDATA(outputBuffer->data),
+										PGLZ_strategy_always);
+	if (compressedByteCount >= 0)
+	{
+		CSTORE_COMPRESS_SET_RAWSIZE(outputBuffer->data, inputBuffer->len);
+		SET_VARSIZE_COMPRESSED(outputBuffer->data,
+							   compressedByteCount + CSTORE_COMPRESS_HDRSZ);
+		compressionResult = true;
+	}
+#else
+
+	compressionResult = pglz_compress(inputBuffer->data, inputBuffer->len,
+									  CSTORE_COMPRESS_RAWDATA(outputBuffer->data),
+									  PGLZ_strategy_always);
+#endif
+
+	if (compressionResult)
+	{
+		outputBuffer->len = VARSIZE(outputBuffer->data);
+	}
+
+	return compressionResult;
+}
+
+
+/*
+ * DecompressBuffer decompresses the given buffer with the given compression
+ * type. This function returns the buffer as-is when no compression is applied.
+ */
+StringInfo
+DecompressBuffer(StringInfo buffer, CompressionType compressionType)
+{
+	StringInfo decompressedBuffer = NULL;
+
+	Assert(compressionType == COMPRESSION_NONE || compressionType == COMPRESSION_PG_LZ);
+
+	if (compressionType == COMPRESSION_NONE)
+	{
+		/* in case of no compression, return buffer */
+		decompressedBuffer = buffer;
+	}
+	else if (compressionType == COMPRESSION_PG_LZ)
+	{
+		uint32 compressedDataSize = VARSIZE(buffer->data) - CSTORE_COMPRESS_HDRSZ;
+		uint32 decompressedDataSize = CSTORE_COMPRESS_RAWSIZE(buffer->data);
+		char *decompressedData = NULL;
+
+		if (compressedDataSize + CSTORE_COMPRESS_HDRSZ != buffer->len)
+		{
+			ereport(ERROR, (errmsg("cannot decompress the buffer"),
+							errdetail("Expected %u bytes, but received %u bytes",
+									  compressedDataSize, buffer->len)));
+		}
+
+		decompressedData = palloc0(decompressedDataSize);
+
+#if PG_VERSION_NUM >= 90500
+
+		int32 decompressedByteCount = 0;
+
+		decompressedByteCount = pglz_decompress(CSTORE_COMPRESS_RAWDATA(buffer->data),
+												compressedDataSize,
+												decompressedData, decompressedDataSize);
+
+		if (decompressedByteCount < 0)
+		{
+			ereport(ERROR, (errmsg("cannot decompress the buffer"),
```

Comment:
I know here is trickier to add a tests since it is in `if PG 9.5` block, but if possible could you add a test?